orm/model/ormtable/bench_test.go

```diff
@@ -0,0 +1,315 @@
+package ormtable_test
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"google.golang.org/protobuf/proto"
+
+	"github.com/cosmos/cosmos-sdk/orm/internal/testkv"
+	"github.com/cosmos/cosmos-sdk/orm/testing/ormtest"
+
+	dbm "github.com/tendermint/tm-db"
+	"gotest.tools/v3/assert"
+
+	"github.com/cosmos/cosmos-sdk/orm/internal/testpb"
+	"github.com/cosmos/cosmos-sdk/orm/model/ormtable"
+	"github.com/cosmos/cosmos-sdk/orm/types/kv"
+)
+
+var balanceTable testpb.BalanceTable
+
+func init() {
+	table, err := ormtable.Build(ormtable.Options{
+		MessageType: (&testpb.Balance{}).ProtoReflect().Type(),
+	})
+	if err != nil {
+		panic(err)
+	}
+
+	balanceTable, err = testpb.NewBalanceTable(table)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func BenchmarkMemory(b *testing.B) {
+	bench(b, func(tb testing.TB) ormtable.Backend {
+		return ormtest.NewMemoryBackend()
+	})
+}
+
+func BenchmarkLevelDB(b *testing.B) {
+	bench(b, testkv.NewGoLevelDBBackend)
+}
+
+func bench(b *testing.B, newBackend func(testing.TB) ormtable.Backend) {
+	b.Run("insert", func(b *testing.B) {
+		b.StopTimer()
+		ctx := ormtable.WrapContextDefault(newBackend(b))
+		b.StartTimer()
+		benchInsert(b, ctx)
+	})
+	b.Run("update", func(b *testing.B) {
+		b.StopTimer()
+		ctx := ormtable.WrapContextDefault(newBackend(b))
+		benchInsert(b, ctx)
+		b.StartTimer()
+		benchUpdate(b, ctx)
+	})
+	b.Run("get", func(b *testing.B) {
+		b.StopTimer()
+		ctx := ormtable.WrapContextDefault(newBackend(b))
+		benchInsert(b, ctx)
+		b.StartTimer()
+		benchGet(b, ctx)
+	})
+	b.Run("delete", func(b *testing.B) {
+		b.StopTimer()
+		ctx := ormtable.WrapContextDefault(newBackend(b))
+		benchInsert(b, ctx)
+		b.StartTimer()
+		benchDelete(b, ctx)
+	})
+}
+
+func benchInsert(b *testing.B, ctx context.Context) {
+	for i := 0; i < b.N; i++ {
+		assert.NilError(b, balanceTable.Insert(ctx, &testpb.Balance{
+			Address: fmt.Sprintf("acct%d", i),
+			Denom:   "bar",
+			Amount:  10,
+		}))
+	}
+}
+
+func benchUpdate(b *testing.B, ctx context.Context) {
+	for i := 0; i < b.N; i++ {
+		assert.NilError(b, balanceTable.Update(ctx, &testpb.Balance{
+			Address: fmt.Sprintf("acct%d", i),
+			Denom:   "bar",
+			Amount:  11,
+		}))
+	}
+}
+
+func benchGet(b *testing.B, ctx context.Context) {
+	for i := 0; i < b.N; i++ {
+		balance, err := balanceTable.Get(ctx, fmt.Sprintf("acct%d", i), "bar")
+		assert.NilError(b, err)
+		assert.Equal(b, uint64(10), balance.Amount)
+	}
+}
+
+func benchDelete(b *testing.B, ctx context.Context) {
+	for i := 0; i < b.N; i++ {
+		assert.NilError(b, balanceTable.Delete(ctx, &testpb.Balance{
+			Address: fmt.Sprintf("acct%d", i),
+			Denom:   "bar",
+		}))
+	}
+}
```

Comment:
All of these functions operate against a single shared global `balanceTable` value. That means they're not necessarily isolated — running benchmark A can affect the results of running benchmark B. 

The only way to get deterministic, reproducible output from benchmarks (and tests!) is to make sure they're fully isolated from each other. In this case, it's easy to fix: just remove the global variable + func init, and create a fresh BalanceTable in each benchmark function.

Reply:
It looks like this is what you're doing in the "manual" benchmarks, below — great! Just use the same approach up here and you're golden 👍

Reply:
The BalanceTable struct is actually stateless - it only encapsulates encoding rules. But I can initialize it in each benchmark to model best practices

Reply:
> The BalanceTable struct is actually stateless - it only encapsulates encoding rules.

Ah, I see!

(On the off chance this is useful feedback: I expected that a value of a type named Table would be stateful, or at least a [stateless] proxy to a single stateful entity. That conclusion was supported by the method signatures of the Table interface, as there's no signal that they would operate on a target provided by the caller. Shuttling mandatory and mutable parameters through the context is a very subtle and uncommon strategy! It would have been helpful to me if it were made more obvious, e.g. if Table were instead TableEncoding, or if the Backend were an explicit parameter to the methods.)